### ISSUE TITLE
Improving the C++ nullable implementation

### ISSUE THREAD

**Issue by glenfe** [Microsoft/bond#167]
1. [x] **1.** When `BOND_NO_CXX11_RVALUE_REFERENCES`  is not defined, and `arg1` is a forwarding-reference (instead of a const l-value reference) then construction of `T` should use `std::forward<Arg1>(arg1)` instead of `arg1`.
   
   For example:
   
   ```
   template<typename Arg1>
   pointer new_value(detail::no_allocator&, Arg1&& arg1)
   {
       return new T(std::forward<Arg1>(arg1));
   }
   ```
2. [x] **2.** In `destroy()` when the C++11 allocator model is available (e.g. `BOOST_NO_CXX11_ALLOCATOR` is not defined) then do not use `alloc.destroy(_value)` but instead use `std::allocator_traits::destroy` because a valid C++11 allocator is not required to have a `destroy` member.
3. [x] **3.** Either make `new_value()` use `std::allocator_traits::construct` (in C++11 - or the allocator's `construct` in C++03) or make `destroy()` invoke the destructor instead of calling `destroy`. i.e. Be consistent: If placement new is used, then manually invoke the destuctor. If allocator's `construct` is used, then use allocator's `destroy`.
4. [x] **4.** Currently `nullable` is confined to custom allocators whose `pointer` type is a raw pointer type compatible with `T*`. In the C++11 allocator model, this is not required to be true: i.e. you can have valid custom allocators whose `pointer` type is a kind of smart pointer. So the `pointer` member type of `nullable` has to be inferred from the allocator's `pointer` member type (via `std::allocator_traits::pointer` in C++11).
5. [x] **5.** It would be better for the first specialization of `nullable` to store an object of type `std::aligned_storage_t<sizeof(value_type), alignof(value_type)>` (instead of storing an object of type `value_type`) to have a better `make_value()` implementation. i.e. To not be reliant on compiler optimizations like RVO to elide any potential copy or moves and construct that `value_type` in-place with the desired arguments. (Of course `boost::aligned_storage` and `boost::alignment_of` can be used instead for compatibility with older compilers).
6. [x] **6.** In both specializations of `nullable`, leverage the EBO (empty-base optimization) for the allocator. Derive privately from `Allocator` instead of storing it as a member so that for empty allocators (e.g. `std::allocator`) you're not increasing the `sizeof(nullable<T>)` by an extra `alignof(value_type)` bytes or `alignof(pointer)` bytes. You can do this derivation conditionally if you want (e.g. if the `Allocator` type is not `final` and is an empty type).

**Comment by glenfe** [comment]
Edit: Edited formatting and added two new points (5 and 6).

**Comment by glenfe** [comment]
Approval came through; I should be able to submit a pull request for this sometime next week.

**Comment by chwarr** [comment]
There's nothing coming up that would address any of these concerns, so please go ahead. These look like they're all relatively independent, so can you try to keep the individual commits small. Individual pull requests are fine if you want to go that route as well.

**Comment by chwarr** [comment]
Thanks for #174. I've merged it. I also edited the issue to convert things over to check boxes so we can  see what's been done so far.

**Comment by glenfe** [comment]
The way that number 5 is optimally (and minimally) implemented depends on two things:
* Your plans for backward compatibility for older C++ language and Boost library versions. For example, any plans on supporting only C++11 and higher (with only VC15 and higher)?
* Whether you still want current semantics where value_type is always constructed even when `hasvalue` is `false` instead of constructed only in cases where `hasvalue` becomes `true`.

**Comment by chwarr** [comment]
We can't drop Visual C++ 2010 (VC11) or Boost 1.59 support yet.

I'm fine with `value_type` only being constructed if we have something sensible to put there. This makes even more sense if we're only reserving space for it inside the nullable.

I looked through the history of the project--including the original Microsoft-internal implementation--to see whether I could suss out why nullable works this way. It looks like it works this way because the initial implementation was succinct, and it was never radically changed even as new features (e.g., allocators) were added:

```
template<typename T, /* ... */>
class nullable
{
public:
    typedef T           value_type;
    ...

    nullable()
        : m_value(),
          m_hasvalue(false)
    {}

    explicit
    nullable(const value_type& value)
        : m_value(value),
          m_hasvalue(true)
    {}

    ...

private:
    value_type  m_value;
    bool        m_hasvalue;
};
```

At is evolved, it was never changed to avoid constructing `value_type`.
